query/src/main/java/io/keen/client/java/Query.java

```diff
@@ -0,0 +1,416 @@
+package io.keen.client.java;
+
+import java.util.Map;
+import java.util.List;
+import java.util.HashMap;
+import java.util.ArrayList;
+
+/**
+ * Created by claireyoung on 5/18/15.
+ */
+public class Query {
+
+    private QueryType queryType;
+
+    // required for all queries
+    private String eventCollection;
+
+    // required by most queries
+    private String targetProperty;
+
+    // optional
+    private List<Map<String, Object>> filters;
+    private String interval;    // requires timeframe to be set
+    private String timezone;
+    private ArrayList<String> groupBy;
+    private Integer maxAge; // integer greater than 30 seconds: https://keen.io/docs/data-analysis/caching/
+
+    // required by the Percentile query
+    private Double percentile;  // 0-100 with two decimal places of precision for example, 99.99
+
+
+    /**
+     * Constructs the map to pass to the JSON handler, so that the proper required
+     * and optional Query arguments can be sent out to the server.
+     *
+     * @return The JSON object map.
+     */
+    public Map<String, Object> ConstructQueryArgs() {
+
+        Map<String, Object> queryArgs = new HashMap<String, Object>();
+
+        StringBuffer queryString = new StringBuffer();
+        if (null != eventCollection) {
+            queryArgs.put(KeenQueryConstants.EVENT_COLLECTION, eventCollection);
+        }
+
+        if (null != interval) {
+            queryArgs.put(KeenQueryConstants.INTERVAL, interval);
+        }
+
+        if (null != timezone) {
+            queryArgs.put(KeenQueryConstants.TIMEZONE, timezone);
+        }
+
+        if (null != groupBy) {
+            queryArgs.put(KeenQueryConstants.GROUP_BY, groupBy);
+        }
+
+        if (null != maxAge) {
+            queryArgs.put(KeenQueryConstants.MAX_AGE, maxAge);
+        }
+
+        if (null != targetProperty) {
+            queryArgs.put(KeenQueryConstants.TARGET_PROPERTY, targetProperty);
+        }
+
+        if (null != percentile) {
+            queryArgs.put(KeenQueryConstants.PERCENTILE, percentile);
+        }
+
+        if (null != filters && filters.isEmpty() == false) {
+            queryArgs.put(KeenQueryConstants.FILTERS, filters);
+        }
+
+        return queryArgs;
+    }
+
+    /**
+     * @return the query type
+     */
+    public QueryType getQueryType() {
+        return this.queryType;
+    }
+
+    /**
+     * @return whether this query has a GroupBy specified.
+     */
+    public boolean hasGroupBy() {return groupBy != null;}
+
+    /**
+     * @return whether this query has an Interval specified.
+     */
+    public boolean hasInterval() {return interval != null;}
+
+
+    /**
+     * Verifies whether the parameters are valid, based on the input query name.
+     *
+     * @param queryType     The type of the query (in {@link QueryType}).
+     * @return       whether the parameters are valid.
+     */
+    public boolean AreParamsValid(QueryType queryType) {
```

Comment:
Concern: Why is the queryType being passed as an argument here? There's already a queryType member variable, and I would have thought we'd want to use that rather than a (potentially different) argument.